File: rtix/process/tests/test_process.cpp
// Copyright (c) 2025 Velodex Robotics, Inc and RTIX Developers.
// Licensed under Apache-2.0. http://www.apache.org/licenses/LICENSE-2.0

#include <google/protobuf/wrappers.pb.h>
#include <gtest/gtest.h>
#include "rtix/api/common.pb.h"
#include "rtix/core/timer.h"
#include "rtix/ipc/channel_map.h"
#include "rtix/ipc/node.h"
#include "rtix/process/process.h"

using google::protobuf::BoolValue;
using google::protobuf::Int64Value;
using rtix::api::common::Code;
using rtix::api::common::Status;
using rtix::core::Timer;
using rtix::ipc::ChannelMap;
using rtix::ipc::Node;
using rtix::process::InputConfig;
using rtix::process::OutputConfig;
using rtix::process::Process;
using rtix::process::ProcessConfig;

double MAX_PROCESS_TIME_S = 3.0;
bool ACTION_VALUE = true;
int INPUT_VALUE = 567;

ChannelMap load_test_channel_map() {
  std::string yaml_file = "/rtix/rtix/process/tests/channel-map.yaml";
  YAML::Node yaml_node = YAML::LoadFile(yaml_file);
  return ChannelMap::LoadYaml(yaml_node);
}

class TestProcess : public Process<BoolValue> {
 public:
  explicit TestProcess(const Node& node, const ProcessConfig& config)
      : Process<BoolValue>(node, config) {
    SPDLOG_INFO("Created test process");
  }
  virtual ~TestProcess() = default;

  void handleAction(const BoolValue& action) override {
    // Need to sleep to give the fixture node time to receive
    Timer::Sleep(0.2);
    publishStatus(Code::SUCCESS, "Handled");
    setRunning(true);
    SPDLOG_INFO("Handled action");
  }

  void stepOuter() override {
    if (receiveInput("in", input)) {
      SPDLOG_INFO("Received input");
    }

    if (getProcessTimeS() > MAX_PROCESS_TIME_S) {
      stop();
    }
  }

  void stepInner() override {
    publishOutput("out", input);
    SPDLOG_INFO(RTIX_FMT("Published output " << input.value()));
  }

 private:
  Int64Value input{};
};

void test_process_callback(ChannelMap channel_map) {
  EXPECT_TRUE(channel_map.nodes.find("test_node") != channel_map.nodes.end());
  auto test_node = Node(channel_map.getNodeConfig("test_node"));

  // Create the process
  ProcessConfig config;
  config.monitor_rate_s = 0.1;
  config.loop_rate_s = 0.05;
  config.action_channel_id = "action";
  config.status_channel_id = "status";
  config.input_channels["in"] = InputConfig{"x"};
  config.output_channels["out"] = OutputConfig{"y"};
  TestProcess test_process(test_node, config);

  // Run for finite amount of time
  test_process.run();
}

TEST(Process, Process) {
  const auto channel_map = load_test_channel_map();

  std::thread test_process_thread(test_process_callback, channel_map);

  // Access the fixture node after the test process has started
  EXPECT_TRUE(channel_map.nodes.find("fixture_node") !=
              channel_map.nodes.end());
  auto fixture_node = Node(channel_map.getNodeConfig("fixture_node"));

  // Give the thread time to spin up
  Timer::Sleep(0.2);

  // Send the action
  BoolValue action{};
  action.set_value(ACTION_VALUE);
  fixture_node.publisher("action")->send(action);

  // Await status
  Status status{};
  fixture_node.subscriber("status")->recv(status);

  // Send the 'x' channel
  Int64Value message{};
  message.set_value(INPUT_VALUE);
  fixture_node.publisher("x")->send(message);

  // Give the process time to return the input value
  Timer::Sleep(0.2);

  // Receive the 'y' channel
  Int64Value output{};
  fixture_node.subscriber("y")->recv(output);

  test_process_thread.join();

  // These will only be true if the input and action was received
  EXPECT_EQ(status.code(), Code::SUCCESS);
  EXPECT_EQ(output.value(), INPUT_VALUE);
}

File: rtix/process/tests/test_process.py
# Copyright (c) 2025 Velodex Robotics, Inc and RTIX Developers.
# Licensed under Apache-2.0. http://www.apache.org/licenses/LICENSE-2.0

import time
import logging
from google.protobuf.wrappers_pb2 import Int64Value, BoolValue

from rtix.types import common_pb2
from rtix.ipc.node import Node
from rtix.core.status import Status
from rtix.process.process import Process, ProcessConfig, InputConfig, OutputConfig

from rtix.ipc.tests.test_channel_map import load_test_channel_map
from rtix.ipc.tests.test_node import ThreadWithAssertion

PROCESS_TEST_YAML_FILE = "/rtix/rtix/process/tests/channel-map.yaml"
MAX_PROCESS_TIME_S = 3.0
ACTION_VALUE = True
INPUT_VALUE = 567


class TestProcess(Process):
    __test__ = False

    def __init__(self, node, config, action):
        super().__init__(node, config, action)
        self.input = Int64Value()
        logging.info("Created test process")

    def handleAction(self, action: BoolValue):
        # Need to sleep to give the fixture node time to receive
        time.sleep(0.2)
        self.publishStatus(status=Status.SUCCESS, msg="Handled")
        self.setRunning(True)
        logging.info("Handled action")

    def stepOuter(self):
        if self.receiveInput(key="in", msg=self.input):
            logging.info("Received input")

        if self.getProcessTimeS() > MAX_PROCESS_TIME_S:
            self.stop()

    def stepInner(self):
        if self.input is not None:
            self.publishOutput(key="out", msg=self.input)
            logging.info(f"Published output {self.input.value}")


def test_process():
    # Assertions within a thread won't automatically fail the pytest test
    # because they are raised in a different context. To correctly capture and
    # propagate these failures, a mechanism is needed to re-raise exceptions in
    # the main thread.  We subclass Thread to raise an exception on join, and
    # pass the thread callbacks in the main loop below.

    channel_map = load_test_channel_map(yaml_file=PROCESS_TEST_YAML_FILE)

    def test_process_callback(channel_map):
        assert "test_node" in channel_map.nodes
        test_node = Node(config=channel_map.nodes["test_node"])

        # Create the process
        test_process = TestProcess(
            node=test_node,
            config=ProcessConfig(
                monitor_rate_s=0.1,
                loop_rate_s=0.05,
                action_channel_id="action",
                status_channel_id="status",
                input_channels={"in": InputConfig(channel_id="x")},
                output_channels={"out": OutputConfig(channel_id="y")},
            ),
            action=BoolValue(),
        )

        # Run for finite amount of time
        test_process.run()

    # Start the thread
    test_process_thread = ThreadWithAssertion(
        target=test_process_callback,
        channel_map=channel_map,
    )

    test_process_thread.start()

    # Access the fixture node after the test process has started
    assert "fixture_node" in channel_map.nodes
    fixture_node = Node(config=channel_map.nodes["fixture_node"])

    # Give the thread time to spin up
    time.sleep(0.2)

    # Send the action
    fixture_node.publisher(channel_id="action").send(msg=BoolValue(
        value=ACTION_VALUE))

    # Await status
    outcome = common_pb2.Outcome()
    fixture_node.subscriber(channel_id="status").recv(msg=outcome)

    # Send the 'x' channel
    fixture_node.publisher(channel_id="x").send(msg=Int64Value(
        value=INPUT_VALUE))

    # Give the process time to return the input value
    time.sleep(0.2)

    # Receive the 'y' channel
    output = Int64Value()
    fixture_node.subscriber(channel_id="y").recv(msg=output)

    test_process_thread.join()

    # These will only be true if the input and action was received
    assert outcome.status == common_pb2.Status.SUCCESS
    assert output.value == INPUT_VALUE


if __name__ == "__main__":
    test_process()
